### PR TITLE
Massive improvements to the install script

### PR DESCRIPTION
**Pull request by ehsan18t**
Hello, I hope you are doing well. First of all, thanks to upload this amazing cursor.

Now, I've made some improvements to the install script. The improvements listed below,
 - Fixed the bug which caused install failure sometimes
 - Fixed some potential issues if user has name with space (eg. `ABC DEF`)
 - Added warnings if not run as admin (installation process won't continue)
 - Removed git dependencies (all done with powershell and cmd)
 - Removed redundant codes (now only 123 lines)
 - Auto open instructions after install
 - Auto cleanup after install
 - Restrict invalid input (invalid input will get reject automatically)

I could make the outputs visually better but I wanted to keep your signatures in the script that why I didn't change existing text and codes if not necessary. Please let me know if I can do more.
Thanks

### CODE REVIEW

**Comment by KiritoKing** [comment]
Hey bro, have you ever tried the possibility to auto change system cursor path with registry editing scripts?

**Comment by KiritoKing** [comment]
> Hey bro, have you ever tried the possibility to auto change system cursor path with registry editing scripts?

I found this might be helpful, but I'm not familiar with cmd scripts.

Here's my screenshot, and the path is `HKEY_CURRENT_USER\Control Panel\Cursors`

image

**Comment by ehsan18t** [comment]
@KiritoKing 
> Hey bro, have you ever tried the possibility to auto change system cursor path with registry editing scripts?

Yes, I've done some research on this topic after add this pr. But the repo woner doesn't seems to be interested in improving the script **considering the fact that 99.999999999% people won't have any issue with the `install.inf`**. That's why I didn't invest any of my time here either.

File: Install_Alt_Run-As-Admin.cmd [hunk 1]
```diff
@@ -1,50 +1,45 @@
...
 echo.
-cd "%userprofile%\Desktop"
 echo.
-mkdir W11-CC-V2.2-HDPI-by-Jepri
+SET "downloaded_cursor=%desktop%\W11-CC-V2.2-HDPI-by-Jepri"
+SET "cursor_zip=W11-CC-V2.2-HDPI-by-Jepri.zip"
+mkdir "%downloaded_cursor%"
 echo.
-curl -LJO  https://github.com/PSGitHubUser1/Windows-11-Cursor-Concept-Pro-v2.x/releases/download/v2.2pro_NEW/W11-CC-V2.2-HDPI-by-Jepri.zip
+SET "DL_Link=https://github.com/PSGitHubUser1/Windows-11-Cursor-Concept-Pro-v2.x/releases/download/v2.2pro_NEW/%cursor_zip%"
```
Comment (PSGitHubUser1) [inline Install_Alt_Run-As-Admin.cmd:35]: the 'DL_Link' is already used before in Instructions.txt....

File: Install_Alt_Run-As-Admin.cmd [hunk 2]
```diff
@@ -1,50 +1,45 @@
...
+SET "DL_Location=%desktop%\%cursor_zip%"
```
Comment (PSGitHubUser1) [inline Install_Alt_Run-As-Admin.cmd:36]: Its also used before.. "DL_Location"..

Maybe it works but I do not want to take risk so..

File: Install_Alt_Run-As-Admin.cmd [hunk 3]
```diff
@@ -1,50 +1,45 @@
...
+Call :Any_Downloader
```
Comment (PSGitHubUser1) [inline Install_Alt_Run-As-Admin.cmd:37]: What's Any_Downloader..

I've never seen this before..?

File: Install_Alt_Run-As-Admin.cmd [hunk 4]
```diff
@@ -1,50 +1,45 @@
...
 echo.
-powershell -command "Expand-Archive -Path %userprofile%\Desktop\W11-CC-V2.2-HDPI-by-Jepri.zip -DestinationPath %userprofile%\Desktop\W11-CC-V2.2-HDPI-by-Jepri"
+powershell -command "Expand-Archive -Path '%desktop%\%cursor_zip%' -DestinationPath '%downloaded_cursor%'"
```
Comment (PSGitHubUser1) [inline Install_Alt_Run-As-Admin.cmd:39]: %desktop% variable is not created.. i guess??

(normally %userprofile% works but %desktop% doesn't, you need to create a variable for this)
make it like umm...?
set "desk=%userprofile%\Desktop"

.......

File: Install_Alt_Run-As-Admin.cmd [hunk 5]
```diff
@@ -53,199 +48,75 @@ echo       2. Dark color
...
-copy "%userprofile%\Downloads\Windows-11-Cursor-Concept-Pro-v2.x\Windows.Cursor.Concept.v2.2.new\light\regular\base\alternate.cur" C:\Windows\Cursors\Win11_Light_HD_Cursors
-echo.
-copy "%userprofile%\Downloads\Windows-11-Cursor-Concept-Pro-v2.x\Windows.Cursor.Concept.v2.2.new\light\regular\base\beam.cur" C:\Windows\Cursors\Win11_Light_HD_Cursors
-echo.
-copy "%userprofile%\Downloads\Windows-11-Cursor-Concept-Pro-v2.x\Windows.Cursor.Concept.v2.2.new\light\regular\01_default\busy.ani" C:\Windows\Cursors\Win11_Light_HD_Cursors
-echo.
-copy "%userprofile%\Downloads\Windows-11-Cursor-Concept-Pro-v2.x\Windows.Cursor.Concept.v2.2.new\light\regular\base\dgn1.cur" C:\Windows\Cursors\Win11_Light_HD_Cursors
-echo.
+@REM Installation Process
+mkdir "%WinDir%\Cursors\Win11_%mode%_HD_Cursors"
```
Comment (PSGitHubUser1) [inline Install_Alt_Run-As-Admin.cmd:59]: What's %mode% ?

its something new to me.. hmmm...

File: Install_Alt_Run-As-Admin.cmd [hunk 6]
```diff
@@ -53,199 +48,75 @@ echo       2. Dark color
...
- echo.
- echo     Invalid option selected. Please choose either 1 or 2.
- echo.
-)
-echo.
-pause
+
+:Any_Downloader
+PowerShell -Command ^
+$ProgressPreference = 'SilentlyContinue';^
```
Comment (PSGitHubUser1) [inline Install_Alt_Run-As-Admin.cmd:88]: omg! You are a powershell expert ?
I cant understand anything here.. 
powershell always feels me like its very complicated thingy.. which my brain cant understand.

File: Install_Alt_Run-As-Admin.cmd [hunk 7]
```diff
@@ -53,199 +48,75 @@ echo       2. Dark color
...
+$dlLink = \"%DL_Link%\";^
+$dlLocation = \"%DL_Location%\";^
+function downloadFile($url, $targetFile)^
+{^
+    $uri = New-Object \"System.Uri\" \"$url\";^
```
Comment (PSGitHubUser1) [inline Install_Alt_Run-As-Admin.cmd:93]: whats uri

File: Install_Alt_Run-As-Admin.cmd [hunk 8]
```diff
@@ -53,199 +48,75 @@ echo       2. Dark color
...
+        $targetStream.Write($buffer, 0, $count);^
+        $count = $responseStream.Read($buffer,0,$buffer.length);^
+        $downloadedBytes = $downloadedBytes + $count;^
+    }^
+    $targetStream.Flush();^
+    $targetStream.Close();^
+    $targetStream.Dispose();^
+    $responseStream.Dispose();^
+}^
+downloadFile $dlLink $dlLocation;
```
Comment (PSGitHubUser1) [inline Install_Alt_Run-As-Admin.cmd:116]: able to understand like 0.0001% things 

btw reply me on discord: infoytech

File: Install_Alt_Run-As-Admin.cmd [hunk 9]
```diff
@@ -53,199 +48,75 @@ echo       2. Dark color
...
+
+ECHO.
+EXIT /B
 
 
-:: THIS FILE IS STILL BEING CONSTRUCTED! (PRE-RELEASE)
+@REM :: THIS FILE IS STILL BEING CONSTRUCTED! (PRE-RELEASE)
```
Comment (PSGitHubUser1) [inline Install_Alt_Run-As-Admin.cmd:122]: it your code works after the changes i suggested (which can be wrong suggestions im not expert programmer) then I will give you credit and remove the banner of "Pre-release" to "Released".

File: Install_Alt_Run-As-Admin.cmd [hunk 10]
```diff
@@ -53,199 +48,75 @@ echo       2. Dark color
...
+    }^
+    $targetStream.Flush();^
+    $targetStream.Close();^
+    $targetStream.Dispose();^
+    $responseStream.Dispose();^
+}^
+downloadFile $dlLink $dlLocation;
+
+ECHO.
+EXIT /B
```
Comment (PSGitHubUser1) [inline Install_Alt_Run-As-Admin.cmd:119]: hmmm

**Review by PSGitHubUser1 — CHANGES_REQUESTED** [review]

**Comment by PSGitHubUser1** [comment]
> Hey bro, have you ever tried the possibility to auto change system cursor path with registry editing scripts?

what do you mean.. (sorrry pls explain me ahemm.. ?)

File: Install_Alt_Run-As-Admin.cmd (hunk 1 above)
Comment (ehsan18t) [reply Install_Alt_Run-As-Admin.cmd:35]: I don't even know where to start the explaination. I don't really want to write a long texts but seems like I don't have a choice. There are many place where re-use/re-declaration are fine. But in this case **this is the only option**. You may/may not already know Batch Script doesn't support function. Right?

But we can write codeblocks similar to function, except it doesn't allow to pass parameters. So what other option do we have?
Lets find out with an example,

```batch
ECHO Doing something...
ECHO Again doing something...
ECHO Another some work...


ECHO Doing something...
ECHO Again doing something...
ECHO Another some work...


ECHO Doing something...
ECHO Again doing something...
ECHO Another some work...
```


Now as you can see we are repeating the codes which is not ideal, right?
we can reduce the redundancy by using code blocks.

```batch
CALL :Work
CALL :Work
CALL :Work

:Work
ECHO Doing something...
ECHO Again doing something...
ECHO Another some work...
EXIT /B
```

Much nicer to look and easier to debug if anything goes wrong. 
The thing that you pointed out is basically for a download function. As I've mentioned before parameter is not supported in batch we can declare a variable which will be used in the code block and we will set the URL and PATH just before we call the function. That's exactly what's been happening here. 

```batch
SET "var=X"
CALL :Work

SET "var=Y"
CALL :Work

SET "var=Z"
CALL :Work

:Work
ECHO Doing something...
ECHO Do something with variable %var%
ECHO Another some work...
EXIT /B
```

Hope this clears a lot of your questions and concenrns.

File: Install_Alt_Run-As-Admin.cmd (hunk 2 above)
Comment (ehsan18t) [reply Install_Alt_Run-As-Admin.cmd:36]: Discussed in the previous question.
>Maybe it works but I do not want to take risk so.

I don't know what risk we are talking about. Do you know?

File: Install_Alt_Run-As-Admin.cmd (hunk 4 above)
Comment (ehsan18t) [reply Install_Alt_Run-As-Admin.cmd:39]: Please read the script. It's declare at `line 11`.

File: Install_Alt_Run-As-Admin.cmd (hunk 6 above)
Comment (ehsan18t) [reply Install_Alt_Run-As-Admin.cmd:88]: Not an expert. But I have done quite some work with ps. Well can't disagree on this one. It's pretty complicated but after working with C/C++, C#, VB, JS..... powershell is not that hard to master......

File: Install_Alt_Run-As-Admin.cmd (hunk 7 above)
Comment (ehsan18t) [reply Install_Alt_Run-As-Admin.cmd:93]: Thats declaration of a powershell local variable.

File: Install_Alt_Run-As-Admin.cmd (hunk 8 above)
Comment (ehsan18t) [reply Install_Alt_Run-As-Admin.cmd:116]: I sure will if I get time for that. I am currently really busy with some paid work also some of my research and projects.

File: Install_Alt_Run-As-Admin.cmd (hunk 9 above)
Comment (ehsan18t) [reply Install_Alt_Run-As-Admin.cmd:122]: Cool. It's already will work 100% times with the 100% of the windows system from XP to most recent 11. Feel free to test.

File: Install_Alt_Run-As-Admin.cmd (hunk 10 above)
Comment (ehsan18t) [reply Install_Alt_Run-As-Admin.cmd:119]: haha thanks. I always try to make codes as simple and reusable as possible.

**Comment by PSGitHubUser1** [comment]
?

File: Install_Alt_Run-As-Admin.cmd (hunk 7 above)
Comment (PSGitHubUser1) [reply Install_Alt_Run-As-Admin.cmd:93]: ooo

**Review by PSGitHubUser1 — APPROVED** [review]

**Comment by PSGitHubUser1** [comment]
what!!!! i approved it but see this

**Comment by ehsan18t** [comment]
ok.give me some time. I will let you know after some debugging.

**Comment by ehsan18t** [comment]
I've found the issue and fixed it. The issue is basically we need to bypass execution policy for powershell to execute any powershell script (if downloaded from internet). I've disabled that in my PC on system level as I need to run many scripts every now and then. Thanks for identifying the issue. It should be fixed by now, also fixed some potential issue that could occur because of old files. Please check and let me know if anyhing else appears.

**Comment by PSGitHubUser1** [comment]
> I've found the issue and fixed it. The issue is basically we need to bypass execution policy for powershell to execute any powershell script (if downloaded from internet). I've disabled that in my PC on system level as I need to run many scripts every now and then. Thanks for identifying the issue. It should be fixed by now, also fixed some potential issue that could occur because of old files. Please check and let me know if anyhing else appears.

The cleanup deletes the instruction file too. Which I don't want.

**Comment by ehsan18t** [comment]
I sure can. I will try to add apply cursor automatically tomorrow. Today I am stopping removing the instructions file.

**Comment by PSGitHubUser1** [comment]
> I sure can. I will try to add apply cursor automatically tomorrow. Today I am stopping removing the instructions file.

When tomorrow will come for you man? I want to merge this PR as fast I can do. I am so excited IDK why.

**Comment by PSGitHubUser1** [comment]
Hey... its still not working idk why. this is only for me or everybody?soo..

```txt

   If your install.inf is not working, you can use this cmd as its alternative. (Windows)



   Downloading the 'Instructions.txt' file... Check your Downloads folder or Desktop.


Import-Module : The specified module 'Terminal-Icons' was not loaded because no valid module file was found in any
module directory.
At C:\Users\rksra\OneDrive\Documents\WindowsPowerShell\Microsoft.PowerShell_profile.ps1:17 char:1
+ Import-Module -Name Terminal-Icons
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (Terminal-Icons:String) [Import-Module], FileNotFoundException
    + FullyQualifiedErrorId : Modules_ModuleNotFound,Microsoft.PowerShell.Commands.ImportModuleCommand

oh-my-posh : The term 'oh-my-posh' is not recognized as the name of a cmdlet, function, script file, or operable
program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At C:\Users\rksra\OneDrive\Documents\WindowsPowerShell\Microsoft.PowerShell_profile.ps1:228 char:1
+ oh-my-posh init pwsh --config ~/jandedobbeleer.omp.json | Invoke-Expr ...
+ ~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (oh-my-posh:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

 >> Downloaded 0K of 0K (∞%) <<


   FOLLOW INSTRUCTIONS GIVEN IN "Instructions.txt"




Import-Module : The specified module 'Terminal-Icons' was not loaded because no valid module file was found in any
module directory.
At C:\Users\rksra\OneDrive\Documents\WindowsPowerShell\Microsoft.PowerShell_profile.ps1:17 char:1
+ Import-Module -Name Terminal-Icons
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (Terminal-Icons:String) [Import-Module], FileNotFoundException
    + FullyQualifiedErrorId : Modules_ModuleNotFound,Microsoft.PowerShell.Commands.ImportModuleCommand
                                                                                                                        oh-my-posh : The term 'oh-my-posh' is not recognized as the name of a cmdlet, function, script file, or operable        program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.      At C:\Users\rksra\OneDrive\Documents\WindowsPowerShell\Microsoft.PowerShell_profile.ps1:228 char:1                      + oh-my-posh init pwsh --config ~/jandedobbeleer.omp.json | Invoke-Expr ...                                             + ~~~~~~~~~~                                                                                                                + CategoryInfo          : ObjectNotFound: (oh-my-posh:String) [], CommandNotFoundException                              + FullyQualifiedErrorId : CommandNotFoundException                                                                                                                                                                                           >> Downloaded 8522K of 25418K (33%) <<                                                                                                                                                                                                         Import-Module : The specified module 'Terminal-Icons' was not loaded because no valid module file was found in any      module directory.                                                                                                       At C:\Users\rksra\OneDrive\Documents\WindowsPowerShell\Microsoft.PowerShell_profile.ps1:17 char:1                       + Import-Module -Name Terminal-Icons
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (Terminal-Icons:String) [Import-Module], FileNotFoundException
    + FullyQualifiedErrorId : Modules_ModuleNotFound,Microsoft.PowerShell.Commands.ImportModuleCommand

oh-my-posh : The term 'oh-my-posh' is not recognized as the name of a cmdlet, function, script file, or operable
program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At C:\Users\rksra\OneDrive\Documents\WindowsPowerShell\Microsoft.PowerShell_profile.ps1:228 char:1
+ oh-my-posh init pwsh --config ~/jandedobbeleer.omp.json | Invoke-Expr ...
+ ~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (oh-my-posh:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

Expand-Archive : The path 'C:\Users\rksra\Desktop\W11-CC-V2.2-HDPI-by-Jepri.zip' either does not exist or is not a
valid file system path.
At line:1 char:1
+ Expand-Archive -Path 'C:\Users\rksra\Desktop\W11-CC-V2.2-HDPI-by-Jepr ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (C:\Users\rksra\...PI-by-Jepri.zip:String) [Expand-Archive], InvalidOpe
   rationException
    + FullyQualifiedErrorId : ArchiveCmdletPathNotFound,Expand-Archive





    Please choose a cursor color:
      1. Light color
      2. Dark color
      3. Exit


Choose a color (only one option to choose, no go back option):
```

**Comment by PSGitHubUser1** [comment]
What happend

**Comment by ehsan18t** [comment]
extremely sorry about the late. I am having some serious health issue. Will try to start working soon. For now, the issue you are having right now is because of **Powershell Theme (Oh-My-Posh)**. I am not sure if I will be able to fix it or not, but I will try.

**Comment by PSGitHubUser1** [comment]
> extremely sorry about the late. I am having some serious health issue. Will try to start working soon. For now, the issue you are having right now is because of **Powershell Theme (Oh-My-Posh)**. I am not sure if I will be able to fix it or not, but I will try. 

Oh ok

**Comment by PSGitHubUser1** [comment]
Well bro leave it as this repo has found this repo...

Read the Readme.md file as you can 'cuz now this repo is going to be deleted.....

**Pull request closed without merging**